test/ExecutionFramework.cpp

```diff
@@ -50,25 +55,294 @@ string getIPCSocketPath()
+		if (!acc)
+			return 0;
+		size_t i = 0;
+		for (; i < _bufferSize && _codeOffset + i < acc->code.size(); i++)
+			_bufferData[i] = acc->code[_codeOffset + i];
+		return i;
+	}
+
+	void selfdestruct(evmc_address const& _addr, evmc_address const& _beneficiary) noexcept final
+	{
+		m_state.accounts[_beneficiary].balance = m_state.accounts[_addr].balance;
+		m_state.accounts.erase(_addr);
+	}
+
+	evmc::result call(evmc_message const& _message) noexcept final
+	{
+		u256 value{convertFromEVMC(_message.value)};
+		Account* sender = account(_message.sender);
+		// TODO are we responsible for checking balance?
+		if (!sender)// || u256(convertFromEVMC(sender->balance)) < value)
+		{
+			evmc_result res{};
+			// TODO correct to use revert here?
+			res.status_code = EVMC_REVERT;
+			return evmc::result{res};
+		}
+
+		State stateBackup = m_state;
+
+		bytes code;
+
+		evmc_message message = _message;
+		if (message.kind == EVMC_CREATE)
+		{
+			// TODO this is not the right formula
+			// TODO is the nonce incremented on failure, too?
+			Address createAddress(keccak256(
+				bytes(begin(message.sender.bytes), end(message.sender.bytes)) +
+				asBytes(to_string(sender->nonce++))
+			));
+			message.destination = convertToEVMC(createAddress);
+			code = bytes(message.input_data, message.input_data + message.input_size);
+		}
+		else if (message.kind == EVMC_DELEGATECALL)
+		{
+			code = m_state.accounts[message.destination].code;
+			message.destination = m_currentAddress;
+		}
+		else if (message.kind == EVMC_CALLCODE)
+		{
+			code = m_state.accounts[message.destination].code;
+			message.destination = m_currentAddress;
+		}
+		else
+			code = m_state.accounts[message.destination].code;
+		//TODO CREATE2
+
+		Account& destination = m_state.accounts[message.destination];
+
+		if (value != 0 && message.kind != EVMC_DELEGATECALL && message.kind != EVMC_CALLCODE)
+		{
+			sender->balance = convertToEVMC(u256(convertFromEVMC(sender->balance)) - value);
+			destination.balance = convertToEVMC(u256(convertFromEVMC(destination.balance)) + value);
+		}
+
+		evmc_address currentAddress = m_currentAddress;
+		m_currentAddress = message.destination;
+		evmc::result result = m_vm.execute(*this, m_evmVersion, message, code.data(), code.size());
+		m_currentAddress = currentAddress;
+
+		if (result.status_code != EVMC_SUCCESS)
+			m_state = stateBackup;
+		else if (message.kind == EVMC_CREATE)
+		{
+			result.create_address = message.destination;
+			destination.code = bytes(result.output_data, result.output_data + result.output_size);
+			destination.codeHash = convertToEVMC(keccak256(destination.code));
+		}
+
+		return result;
+	}
+
+	evmc_tx_context get_tx_context() noexcept final { return {}; }
+
+	evmc_bytes32 get_block_hash(int64_t number) noexcept final
+	{
+		int64_t current_block_number = get_tx_context().block_number;
+
+		auto example_block_hash = evmc_bytes32{};
+		if (number < current_block_number && number >= current_block_number - 256)
```

Comment:
You don't need to perform these checks, the VM has already done it. https://github.com/ethereum/evmone/blob/master/lib/evmone/instructions.cpp#L658-L669

Reply:
This was taken from the example host, you might wan to update it.